### PR TITLE
[Android] Resolve Android-specific active issues in System.Net.Security and System.Security.Cryptography

### LINKED ISSUE #68206: [Android] Some System.Net.Security tests are failing
The following `System.Net.Security` tests are failing on Android:
- CertificateValidationClientServer.CertificateSelectionCallback_DelayedCertificate_OK
- ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails
- SslStreamAlpnTest_Async.SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail
- SslStreamAlpnTest_Sync.SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail
- SslStreamNetworkStreamTest.SslStream_TargetHostName_Succeeds
- SslStreamNetworkStreamTest.SslStream_ServerUntrustedCaWithCustomTrust_OK
- SslStreamNetworkStreamTest.SslStream_ClientCertificate_SendsChain
- SslStreamNetworkStreamTest.SslStream_ClientCertificateContext_SendsChain
- SslStreamSniTest.SslStream_ClientSendsSNIServerReceives_Ok
- SslStreamSniTest.SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection
- SslStreamSniTest.UnencodedHostName_ValidatesCertificate
- TransportContextTest.TransportContext_ConnectToServerWithSsl_GetExpectedChannelBindings

### LINKED ISSUE #100224: [mono][android] Failures in System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_*
## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=615665
Build error leg or test failing: System.Security.Cryptography.Tests

Failing on:
- `android-arm Release AllSubsets_Mono`
- `android-arm64 Release AllSubsets_Mono`

Error log extracted from the device:
```
03-24 15:06:01.820 10844 12116 I DOTNET  : Failed tests:
03-24 15:06:01.820 10844 12116 I DOTNET  : 1) 	[FAIL] System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithSha256RsaSignature   Test name: System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithSha256RsaSignature
03-24 15:06:01.820 10844 12116 I DOTNET  :    Assembly:  [System.Security.Cryptography.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]
03-24 15:06:01.820 10844 12116 I DOTNET  :    Exception messages: Expected: Value PartialChain (of enum type System.Security.Cryptography.X509Certificates.X509ChainStatusFlags) to have the flag UntrustedRoot set.   Exception stack traces:    at System.AssertExtensions.HasFlag[X509ChainStatusFlags](X509ChainStatusFlags expected, X509ChainStatusFlags actual, String userMessage) in /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:line 413
03-24 15:06:01.820 10844 12116 I DOTNET  :    at System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithSha256RsaSignature() in /_/src/libraries/System.Security.Cryptography/tests/X509Certificates/ChainTests.cs:line 1286
03-24 15:06:01.820 10844 12116 I DOTNET  :    at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
03-24 15:06:01.820 10844 12116 I DOTNET  :    at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
03-24 15:06:01.820 10844 12116 I DOTNET  :    Execution time: 0.1970542
03-24 15:06:01.820 10844 12116 I DOTNET  : 2) 	[FAIL] System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithUnknownOidSignature   Test name: System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithUnknownOidSignature
03-24 15:06:01.820 10844 12116 I DOTNET  :    Assembly:  [System.Security.Cryptography.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]
03-24 15:06:01.820 10844 12116 I DOTNET  :    Exception messages: Assert.ThrowsAny() Failure: No exception was thrown
03-24 15:06:01.820 10844 12116 I DOTNET  : Expected: typeof(System.Security.Cryptography.CryptographicException)   Exception stack traces:    at System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_WithUnknownOidSignature() in /_/src/libraries/System.Security.Cryptography/tests/X509Certificates/ChainTests.cs:line 1319
03-24 15:06:01.820 10844 12116 I DOTNET  :    at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
03-24 15:06:01.820 10844 12116 I DOTNET  :    at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
03-24 15:06:01.820 10844 12116 I DOTNET  :    Execution time: 0.1903711
```

### Known Issue Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "[FAIL] System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_",
  "BuildRetry": false,
  "ExcludeConsoleLog": true
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=615665
**Error message validated:** `[[FAIL] System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChainForSelfSignedCertificate_`]
**Result validation:** :x: Known issue did not match with the provided build.
**Validation performed at:** 3/25/2024 8:52:18 AM UTC

#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|0|

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #68206
Closes #100224

This PR resolves Android-specific active issues in System.Net.Security and System.Security.Cryptography

- Android/java has stricter requirements on the hostnames it allows for SNI/ALPN
- The Android implementation of SslStream server mode doesn't handle ALPN as expected
- Some scenarios are only partially supported on Android
- Some scenarios are not supported on Android
- The OCSP protocol uses plaintext HTTP and our app setup didn't allow that. This requires a change in the app manifest.

/cc @wfurt

### CODE REVIEW

File: src/libraries/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs [hunk 1]
```diff
@@ -60,7 +59,16 @@ public async Task ServerAsyncAuthenticate_EachSupportedProtocol_Success(SslProto
                 });
 
             Assert.NotNull(e);
-            Assert.IsType<AuthenticationException>(e);
+
+            if (OperatingSystem.IsAndroid() && (PlatformDetection.IsX64Process || PlatformDetection.IsX86Process))
+            {
+                // On Android running on x64 or x86 the server side sometimes throws IOException instead of AuthenticationException
+                Assert.True(typeof(IOException) == e.GetType() || typeof(AuthenticationException) == e.GetType(), $"Unexpected exception type: {e.GetType()}");
```
Comment (wfurt) [inline src/libraries/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs:66]: `e is IOException || e is AuthenticationException` ?
Also this seems just like timing, right? I'm wondering if you really need the architecture check - I would keep it simple.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 2]
```diff
@@ -695,10 +695,15 @@ public async Task SslStream_NestedAuth_Throws()
         [InlineData(false, true)]
         [InlineData(false, false)]
         [InlineData(true, true)]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/68206", TestPlatforms.Android)]
         public async Task SslStream_TargetHostName_Succeeds(bool useEmptyName, bool useCallback)
         {
             string targetName = useEmptyName ? string.Empty : Guid.NewGuid().ToString("N");
+            if (PlatformDetection.IsAndroid && !useEmptyName)
```
Comment (wfurt) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:701]: I would probably use something like `$GUID.test.org` or `$GUID.SslStream_TargetHostName_Succeeds.net` or   use `localhost` on all platforms. Using GUID makes it sometimes easier to debug e.g. you can see unique SNI in packet capture and it avoids side effects with TLS resume and credential cache. Other than that the actual string does not matter AFAIK.

File: src/libraries/System.Net.Security/tests/FunctionalTests/TransportContextTest.cs [hunk 3]
```diff
@@ -50,6 +49,10 @@ private static void CheckTransportContext(TransportContext context)
             {
                 Assert.True(cbt2 == null, "ChannelBindingKind.Unique token data is not expected on OSX platform.");
             }
+            else if (OperatingSystem.IsAndroid())
```
Comment (wfurt) [inline src/libraries/System.Net.Security/tests/FunctionalTests/TransportContextTest.cs:52]: I would probably merge it with the OSX branch above and remove platform from the message or use interpolated string. We may get more in the future.

**Review by wfurt — APPROVED** [review]
I left few comments. Generally looks OK to me.

**Comment by simonrozsival** [comment]
The failing tests in runtime-extra-platforms are unrelated to this PR

**Pull request merged**